### PR TITLE
Add Gridsearch code scripts and remove old codes that are now invalid

### PR DESCRIPTION
**Pull request by chaithyagr**
This fixes the issue #69 by defining a new methodology to define arguments to a script which later iterates through cases.
Adding Zac right now for review of codes, however, I am thinking of getting at least one travis test up to ensure sanity in the meantime.

### CODE REVIEW

File: mri/tests/test_scripts.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    def setUp(self):
+        """ Setup the test.
+        """
+        self.N = 64
+        self.max_iter = 10
+        self.n_jobs = 2
+        self.num_channels = 2
+
+    def test_gridsearch_single_channel(self):
+        """Test Gridsearch script in mri.scripts for
```
Comment (chaithyagr) [inline mri/tests/test_scripts.py]: @zaccharieramzi I am just testing sanity for codes for now, I am not sure what more we can test with gridsearch, any ideas?

Comment (philouc) [reply mri/tests/test_scripts.py]: Coud we define a prescribed set of values for a given parameter to quickly test the grid search functionality?

Comment (chaithyagr) [reply mri/tests/test_scripts.py]: Hmm you mean we know what is the appropriate `mu` and see if we obtain the right value of `mu`? 
For now, the code is general in the sense that the user gets to define what criteria to use for comparing results, I need to still work on result selection aspect, but surely this can be done

Comment (philouc) [reply mri/tests/test_scripts.py]: Not necessarily, here this idea would be to just test that you can correctly iterate over multiple values for the parameter of interest and return the value that gives the best score, whatever the metric used (SSIM, pSNR, etc).

File: examples/gridsearch_launch.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
+"""
+Neuroimaging cartesian reconstruction
+=====================================
+
+Author: Chaithya G R
+
+In this tutorial we will use the pysap-mri's launch grid helper function
+to carry out grid search. We will search for best regulatization weight
```
Comment (zaccharieramzi) [inline examples/gridsearch_launch.py]: typo regularisation (or with a z I don't care)

File: examples/gridsearch_launch.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+reconstruction and best wavelet for reconstruction, while search for a
```
Comment (zaccharieramzi) [inline examples/gridsearch_launch.py]: I think the first reconstruction is not supposed to be here

Comment (zaccharieramzi) [reply examples/gridsearch_launch.py]: What do you mean by:
> while search for a space of 5 weight values

File: examples/gridsearch_launch.py (hunk 2 above)
Comment (chaithyagr) [reply examples/gridsearch_launch.py]: Cool

File: examples/gridsearch_launch.py (hunk 3 above)
Comment (chaithyagr) [reply examples/gridsearch_launch.py]: Ahh, that's bad English fixing. What I wanted to say was that we search for best value of `mu` searched in 5 values in a logarithmic grid

File: mri/scripts/gridsearch.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    init_classes = [
+        all_reformatted_kwargs[i]['init_class']
+        for i in range(len(all_reformatted_kwargs)-1)
+    ]
+    # Obtain key names for all cases
+    key_names = list([
+        list(all_reformatted_kwargs[i]['args'].keys())
+        for i in range(len(all_reformatted_kwargs))
+    ])
+    reshaped_cross_product = []
```
Comment (zaccharieramzi) [inline mri/scripts/gridsearch.py]: I am not sure I understand what's happening here.

File: mri/tests/test_scripts.py (hunk 1 above)
Comment (chaithyagr) [reply mri/tests/test_scripts.py]: Seems like a good idea, let me do this and update

File: mri/scripts/gridsearch.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+    Parameters
+    ----------
+    case: list of parameters
+        keyword arguments for different operators for reconstruction
+    key_names: list of strings
+        names of keyword arguments specified for reconstruction in 'case'
+    init_classes: list of classes
+        classes for varoius operators. must be in format
+        [linear, regularizer, reconstructor]
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: I have been thinking about this, the other alternative is to provide strings that specify if it is cartesian/non-cartesian, what kind of reconstruction etc, and thereby map them, but code maintenance wise its too much of overhead.

File: mri/scripts/gridsearch.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    kspace_data: np.ndarray
+        the kspace data for reconstruction
+    fourier_kwargs: dict
+        holds dictionary with init_class pointing to fourier class to be used
+        and args having keyword arguments for initialization
+        NOTE: We declare fourier operator inside this function to allow
+        parallel execution as NonCartesianFFT cannot be pickled.
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Since NonCartesianFFT, particularly NFFT uses cython, it had pickling errors:

`TypeError: no default __reduce__ due to non-trivial __cinit__`

However, there are some solutions I could find https://stackoverflow.com/questions/12646436/pickle-cython-class, but I didnt give it a try as they mostly would br PRs to pyNFFT

File: mri/scripts/gridsearch.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        fourier_op=fourier_op,
+        linear_op=linear_op,
+        regularizer_op=regularizer_op,
+        **dict(zip(key_names[2], case[2]))
+    )
+    raw_results = reconstructor.reconstruct(
+        kspace_data=kspace_data,
+        **dict(zip(key_names[3], case[3])),
+    )
+    return raw_results, case
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Yeah now that I think of it, we already have it, my bad. :man_facepalming:

File: mri/scripts/gridsearch.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+
+def gather_result(metric, metric_direction, results):
+    """ Gather the best reconstruction result.
+
+    Parameters:
+    -----------
+    metric: str,
+        the name of the metric, it will become a dict key in the output dict.
+    metric_direction: bool,
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: But the default depends on passing the metric itself, which I am not sure if user will always do.
Note that even for SSIM calculation, sometimes, the metric would be more intricate as our functions sometimes return alpha and we may have to map things differently

File: mri/scripts/gridsearch.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    results: list of list,
+        list of the raw results of the gridsearch
+
+    Return:
+    -------
+    results and location of best results in given set of raw results
+    """
+    list_metric = []
+    for res in results:
+        list_metric.append(res[2][metric]['values'][-1])
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Well, it basically is as follows:
raw_results is : `x_final, costs, metrics`

Now we can obtain individually and also work through

File: mri/scripts/gridsearch.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+    list_metric = np.array(list_metric)
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: makes sense

File: mri/scripts/gridsearch.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        best_metric = list_metric.max()
+        best_idx = list_metric.argmax()
+    else:
+        best_metric = list_metric.min()
+        best_idx = list_metric.argmin()
+
+    return best_metric, best_idx
+
+
+def launch_grid(linear_kwargs, regularizer_kwargs, reconstructor_kwargs,
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Hmm, will get to this. I will have to think of some sensible combination of defaults

File: mri/scripts/gridsearch.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        dictionary for linear operator parameters
+    regularizer_kwargs: dict
+        dictionary for regularizer operator parameters
+    reconstructor_kwargs: dict
+        dictionary for reconstructor operator parameters
+    optimizer_kwargs: dict
+        dictionary for optimizer key word arguments
+    compare_metric_details: dict default None
+        dictionary that holds the metric to be compared and metric
+        direction please refer to `gather_result` documentation.
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: We just get all the results and it is upto the user to choose the best result based on his/her interprestation and comparision. This is the case when people dont want to use metrics

File: mri/scripts/gridsearch.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    sets up right parameters for '_reconstruct_case' function.
+    Please check the example code for more details.
+
+    Parameters
+    ----------
+    linear_kwargs: dict
+        dictionary for linear operator parameters
+    regularizer_kwargs: dict
+        dictionary for regularizer operator parameters
+    reconstructor_kwargs: dict
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Well, it is just to define what parameters are needed. We need to define reconstructor within the function, so similar to Fourier_op, we send the class and constant keyword arguments here

File: mri/scripts/gridsearch.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+        these holds extra arguments that are passed to
+        '_reconstruct_case' function.
+        important arguments are 'ksapce_data' and 'fourier_kwargs'
+    """
+    # Convert non-list elements to list so that we can create
+    # search space
+    all_reformatted_kwargs = []
+    for specific_kwargs in [linear_kwargs, regularizer_kwargs,
+                            reconstructor_kwargs, optimizer_kwargs]:
+        for key, value in specific_kwargs['args'].items():
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Yes, changing.
I am calling all dictionaries as `_params` and later replacing `args` with `kwargs`

File: mri/scripts/gridsearch.py [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+            if not isinstance(value, list) and \
+                    not isinstance(value, np.ndarray):
+                specific_kwargs['args'][key] = [value]
+        all_reformatted_kwargs.append(specific_kwargs)
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: It was not in place as I expected, so I needed to do it :cry:

File: examples/gridsearch_launch.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+            'weights': np.geomspace(1e-8, 1e-6, 5),
+        }
+}
+optimizer_kwargs = {
+    # Just following convention
+    'args':
+        {
+            'optimization_alg': 'fista',
+            'num_iterations': 20,
+            'metrics': metrics,
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply examples/gridsearch_launch.py]: The metrics? I dont think that is wise, there are many corner cases to handle, let me see how we can proceed with this

File: mri/scripts/gridsearch.py (hunk 4 above)
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: It is just reshaping the arguments same as key_names, this makes the _reconstruct_case function more readable

File: mri/scripts/gridsearch.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+        '_reconstruct_case' function.
+        important arguments are 'ksapce_data' and 'fourier_kwargs'
+    """
+    # Convert non-list elements to list so that we can create
+    # search space
+    all_reformatted_kwargs = []
+    for specific_kwargs in [linear_kwargs, regularizer_kwargs,
+                            reconstructor_kwargs, optimizer_kwargs]:
+        for key, value in specific_kwargs['args'].items():
+            if not isinstance(value, list) and \
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: I didnt consider this as strings and dictionaries are considered `Iterable`, your check gives a True even for strings, dictionaries etc.
I will need a more complicated logical statement here in that case

File: mri/scripts/gridsearch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: One advantage is that this init_classes is created by the programmer and not the user, so I think this should be fine, for what its worth, this is an internal function and user cannot access this, do you still find possible errors? 

Another option is to pass all classes individual, which bloats up the number of parameters

File: mri/scripts/gridsearch.py [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+    """Internal Function to carry out reconstruction for a
+    special case. This function pulls in appropriate keyword arguments
+    from input and declares appropriate Linear, Fourier and Regularizer
+    Operators. These operators later are used to create the image model
+    by defining the reconstructor. Then the reconstruction is carried and
+    results are returned
+
+    Parameters
+    ----------
+    case: list of parameters
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: changed

File: mri/scripts/gridsearch.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: changed

File: mri/scripts/gridsearch.py [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+    # search space
+    all_reformatted_kwargs = []
+    for specific_kwargs in [linear_kwargs, regularizer_kwargs,
+                            reconstructor_kwargs, optimizer_kwargs]:
+        for key, value in specific_kwargs['args'].items():
+            if not isinstance(value, list) and \
+                    not isinstance(value, np.ndarray):
+                specific_kwargs['args'][key] = [value]
+        all_reformatted_kwargs.append(specific_kwargs)
+    # Create Search space
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Reformatted most codes here as we already have loop above, moved most helper variable creations to there

File: mri/tests/test_scripts.py [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+        single channel reconstruction this is a test of sanity
+        and not if the reconstruction is right.
+        """
+        image = get_sample_data('2d-mri')
+        mask = get_sample_data('cartesian-mri-mask')
+        kspace_loc = convert_mask_to_locations(mask.data)
+        fourier_op = FFT(samples=kspace_loc, shape=image.shape)
```
Comment (zaccharieramzi) [inline mri/tests/test_scripts.py]: if you undersample, there is no guarantee that 0 will be the best regularisation weight.

File: mri/scripts/gridsearch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: I think I would be happier to have a higher number of parameters. 
I understand this is a programmer-used function, but still, let's make our lives easier.

If you have a lot of parameters, you can always use `kwargs` when calling the function to make sure not to mix anything up.

It might also be an indicator of bad design. For example, couldn't we imagine instantiating all the objects outside the parallel call and passing them to the job. This might also resolve the whole NFFT not being pickled stuff no?

EDIT
------

Actually sorry got mixed up, the last comment is shit, the NFFT problem was actually the other way around.

File: mri/scripts/gridsearch.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: couldn't we have a switch instead ? this is really cumbersome, and the user shouldn't have to define the fourier op twice, which right now is the case.

File: mri/scripts/gridsearch.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: I mean for the `metric_direction`, if you receive a `metric` which if lowered contains ssim, psnr or accuracy, then the direction is `True`. That has just to do with the behaviour of this function nothing else.

File: mri/scripts/gridsearch.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: True ! 
In this case let's just make this shorter (and slightly more general) with:
```python
if not isinstance(value, (tuple, list, np.ndarray)):
```

File: mri/scripts/gridsearch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: And the "parameter bloating" is also true for the `grid_search_test_case` param.

File: mri/scripts/gridsearch.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: I see, well fair

File: mri/scripts/gridsearch.py [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+        if True the higher the better the metric value is (like for `ssim`),
+        else the lower the better.
+    results: list of list,
+        list of the raw results of the gridsearch
+
+    Return:
+    -------
+    results and location of best results in given set of raw results
+    """
+    list_metric = np.array([res[2][metric]['values'][-1]
```
Comment (zaccharieramzi) [inline mri/scripts/gridsearch.py]: nitpicking, but always try to make sure the multi-line statements are not going to cause potential unnecessary changes in the future.

File: mri/scripts/gridsearch.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: yes but we should avoid that, we should have the reconstructor passed to the function once and that's it.

File: mri/scripts/gridsearch.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: I am very surprised about this.
I tried the following:
```python
linear_params = { 
   'init_class': WaveletN, 
   'kwargs': 
       { 
           'wavelet_name': ['sym8', 'sym12'], 
           'nb_scale': [3, 4],
           'linear_params': 4,
       } 
}   

for specific_params in [linear_params]: 
    for key, value in specific_params['kwargs'].items(): 
        if not isinstance(value, (tuple, list, np.ndarray)): 
            specific_params['kwargs'][key] = [value] 

print(linear_params)
```

And it printed:
```
{'init_class': 1,
 'kwargs': {'wavelet_name': ['sym8', 'sym12'],
  'nb_scale': [3, 4],
  'single_value': [4]}}
```

So this little example shows me that it should be inline.

File: mri/scripts/gridsearch.py [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+    for specific_params in [linear_params, regularizer_params,
+                            reconstructor_params, optimizer_params]:
+        for key, value in specific_params['kwargs'].items():
+            if not isinstance(value, list) and \
+                    not isinstance(value, np.ndarray):
+                specific_params['kwargs'][key] = [value]
+        all_reformatted_params.append(
+            list(specific_params['kwargs'].values()))
+        # Obtain Initialization classes
+        if specific_params != optimizer_params:
```
Comment (zaccharieramzi) [inline mri/scripts/gridsearch.py]: have you checked that this works even when `optimizer_params` has a non-list value?

File: mri/scripts/gridsearch.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: I don't understand what you wrote.

File: mri/scripts/gridsearch.py (hunk 4 above)
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: can you detail a bit this part, like have it in several lines with variable names that explain what's happening (with comments potentially).

File: mri/scripts/gridsearch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: >And the "parameter bloating" is also true for the grid_search_test_case param.

I dont get you here, so you want to bloat up this too? Like regularizer_params, optimizer_params etc? 
If this is True, maybe we can combine Initclasses and params and pass them in dictionary? (Similar to how user provides it, just that they hold just the parameters and not of list for searching)

File: mri/scripts/gridsearch.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: What do you mean by switch? 
Even if we used a swtich statement to define NFFT or Cartesian, we would still need to pass all the parameters twice right?

File: mri/scripts/gridsearch.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: changed

File: mri/scripts/gridsearch.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: I mean to say that I got rid of that statement, and used earlier loop in code

File: mri/scripts/gridsearch.py (hunk 4 above)
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Done, However, I think I just added more code, let me know if it can be made better

File: mri/tests/test_scripts.py (hunk 20 above)
Comment (chaithyagr) [reply mri/tests/test_scripts.py]: Oh yeah, fixed this and also added a check to ensure that reconstructed image is same as IFFT

File: mri/scripts/gridsearch.py (hunk 21 above)
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Fixed

File: mri/scripts/gridsearch.py (hunk 22 above)
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: In test we use optimizer_params as:

```python
optimizer_params = {
    'kwargs':{
         'optimization_alg': 'fista'
         'num_iterations': 10,
         'metrics': metrics,
    }
}
```

And above works, I dont know why would you be worried aboput this specific case

File: mri/scripts/gridsearch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: yes or even better, a test case class holding all of this. This way you don't have string related errors that are not caught by linters.

[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Agreed, let me see how to reformat, this is probably the best way.

File: mri/scripts/gridsearch.py (hunk 22 above)
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: now that it's inplace there is not a problem anymore

File: mri/scripts/gridsearch.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: yes but there are still some instance below

[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Hmm, are you refering to L231? I dont know how to get rid of it there, the length is number of test cases, at best I can split it to:
```python
num_test_cases = len(cross_product_list)
for i in range(num_test_cases)
    ....
```

File: mri/scripts/gridsearch.py (hunk 4 above)
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: why do you need to do `*tuple(linear_params['kwargs'].values())` and not just `*linear_params['kwargs'].values()`?

File: mri/scripts/gridsearch.py (hunk 21 above)
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: here I would argue the most readable way is:
```python
list_metric = np.array([
	res[2][metric]['values'][-1] for res in results
])
```

File: mri/tests/test_scripts.py (hunk 20 above)
Comment (zaccharieramzi) [reply mri/tests/test_scripts.py]: can you add a little comment specifying that, saying that we don't undersample willignly so that the best `mu` is indeed 0.

Comment (chaithyagr) [reply mri/tests/test_scripts.py]: Oh yeah. on it

File: mri/scripts/gridsearch.py (hunk 4 above)
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Oh yeah, tuple is no longer needed.

File: mri/scripts/gridsearch.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Done

File: mri/scripts/gridsearch.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Done

File: mri/scripts/gridsearch.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Based on offline discussion, we really can't do this (given that reconstructor init needs all the operators to be defined), but at least we plan to provide this details in a different way

File: mri/scripts/gridsearch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Done

File: mri/scripts/gridsearch.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Added both fouier_op and fourier_params. Example tests with fourier_op and `FFT` and test with fourier_params and `NonCartesianFFT`.
Also I opened issue in pyNFFT about adding __reduce__ : https://github.com/pyNFFT/pyNFFT/issues/71

**Comment by chaithyagr** [comment]
Hopefully all changes requested are in, Adding Zack for final review and sign off

File: mri/scripts/gridsearch.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: add in the default values we support in the docs

File: mri/scripts/gridsearch.py [hunk 23]
```diff
@@ -0,0 +1,289 @@
...
+    Return:
+    -------
+    results and location of best results in given set of raw results
+    """
+    # Make metric string lower case
+    metric = metric.lower()
+    list_metric = np.array([
+        res[2][metric]['values'][-1] for res in results
+    ])
+    if metric_direction is None:
```
Comment (zaccharieramzi) [inline mri/scripts/gridsearch.py]: add a comment saying what's happening underneath.

Also you could add `'mse'` as a supported metric.

File: mri/scripts/gridsearch.py [hunk 24]
```diff
@@ -0,0 +1,289 @@
...
+    optimizer_params: dict, default None
+        dictionary for optimizer key word arguments
+        if None, a FISTA optimization is done for 100 iterations
+    fourier_params: dict, default None
+            holds dictionary with init_class pointing to fourier
+            class to be used and args having keyword arguments for
+            initialization
+            This is passed only if fourier_op is None so that fourier_op can be
+            made on spot during reconstruction.
+            NOTE: We declare fourier operator inside this function to allow
```
Comment (zaccharieramzi) [inline mri/scripts/gridsearch.py]: I don't think this is what we want to do. This boils down to the same situation as before. You should be able to create this dictionary yourself from the `fourier_op` passed in by the user.
Basically, you would have a test `if isinstance(fourier_op, NFFT): fourier_params = {...`.
(and you would set `fourier_op` to `None`). This way the user doesn't have to "write" the fourier operator twice.

File: mri/scripts/gridsearch.py [hunk 25]
```diff
@@ -0,0 +1,289 @@
...
+            all_kwargs_values.append(param_kwargs)
+        test_cases.append(_TestCase(
+            *init_classes,
+            *all_kwargs_values)
+        )
+
+    # Call for reconstruction
+    results = Parallel(n_jobs=n_jobs)(
+        delayed(test_cases[i].reconstruct_case)(
+            kspace_data=kspace_data,
```
Comment (zaccharieramzi) [inline mri/scripts/gridsearch.py]: can't all this be part of test case? why pass it at run time?

File: mri/scripts/gridsearch.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply mri/scripts/gridsearch.py]: Well in this for example:

```python
for i in range(number_of_test_cases):
    # Iterator for i-th test case
    iterator = iter(cross_product_list[i])
```

First the comment is a bit redundant. Second why not do:
```python
for test_case in cross_product_list:
    iterator = iter(test_case)
```

since you don't use `i` afterwards (and even if you were you could always use `zip` or `enumerate` in last resort).

Same goes for the parallel execution.

File: mri/tests/test_scripts.py [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+
+    def test_gridsearch_single_channel(self):
+        """Test Gridsearch script in mri.scripts for
+        single channel reconstruction this is a test of sanity
+        and not if the reconstruction is right.
+        """
+        image = get_sample_data('2d-mri')
+        mask = np.ones(image.shape)
+        kspace_loc = convert_mask_to_locations(mask)
+        fourier_op = NonCartesianFFT(samples=kspace_loc, shape=image.shape)
```
Comment (zaccharieramzi) [inline mri/tests/test_scripts.py]: why did you change to NFFT?

Comment (chaithyagr) [reply mri/tests/test_scripts.py]: The test tests for NFFT, we have an example to test for FFT. I wanted to span both eventually. Given that travis is long, I didnt want to add both here itself.

File: mri/scripts/gridsearch.py (hunk 24 above)
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Ahh, that makes more sense...

File: mri/scripts/gridsearch.py (hunk 25 above)
Comment (chaithyagr) [reply mri/scripts/gridsearch.py]: Techincally, we can move everything including fourier_op to test_case. 
I dont have any particular reason for this.

Only thing I am trying to do is to create reconstructor class in this function, as that results in iterations to get Lipchitz constant which I want to be parallel

**Comment by chaithyagr** [comment]
Can this be merged @zaccharieramzi ? Anything left?
